drivers/cleanup.py

```diff
@@ -2532,48 +2538,58 @@ def normalizeType(type):
     return type
 
 def _gcLoop(sr, dryRun):
-    failedCandidates = []
-    while True:
-        if not sr.xapi.isPluggedHere():
-            Util.log("SR no longer attached, exiting")
-            break
-        sr.scanLocked()
-        if not sr.hasWork():
-            Util.log("No work, exiting")
-            break
-
-        if not lockRunning.acquireNoblock():
-            Util.log("Another instance already running, exiting")
-            break
-        try:
-            if not sr.gcEnabled():
+    if not lockActive.acquireNoblock():
+        Util.log("Another GC instance already active, exiting")
+        return
+    try:
+        # TODO: make the delay configurable
+        Util.log("GC active, about to go quiet")
+        time.sleep(5 * 60)
+        Util.log("GC active, quiet period ended")
+
+        while True:
+            if not sr.xapi.isPluggedHere():
+                Util.log("SR no longer attached, exiting")
                 break
-            sr.cleanupCoalesceJournals()
             sr.scanLocked()
-            sr.updateBlockInfo()
-
-            if len(sr.findGarbage()) > 0:
-                sr.garbageCollect(dryRun)
-                sr.xapi.srUpdate()
-                continue
+            if not sr.hasWork():
+                Util.log("No work, exiting")
+                break
 
-            candidate = sr.findCoalesceable()
-            if candidate:
-                util.fistpoint.activate("LVHDRT_finding_a_suitable_pair",sr.uuid)
-                sr.coalesce(candidate, dryRun)
-                sr.xapi.srUpdate()
-                continue
+            if not lockRunning.acquireNoblock():
+                Util.log("Unable to acquire GC running lock.")
+                return
+            try:
+                if not sr.gcEnabled():
+                    break
+                sr.cleanupCoalesceJournals()
+                sr.scanLocked()
+                sr.updateBlockInfo()
+
+                howmany = len(sr.findGarbage())
+                if howmany > 0:
+                    Util.log("Found %d orphaned vdis" % howmany)
+                    sr.garbageCollect(dryRun)
+                    sr.xapi.srUpdate()
+
```

Comment:
It's not required and allows the coalesce to start without continually finding garbage. As it happens with the inlining of the deletes in the SR and the coalesce the garbage path is never taken (unless in case of crash during LV delete).